README
NAME

    Yars - Yet Another RESTful-Archive Service

VERSION

    version 1.04

DESCRIPTION

    Yars is a simple RESTful server for data storage.

    It allows files to be PUT and GET based on their md5 sums and
    filenames, and uses a distributed hash table to store the files across
    any number of hosts and disks.

    Files are assigned to disks and hosts based on their md5s in the
    following manner :

    The first N digits of the md5 are considered the "bucket" for a file.
    e.g. for N=2, 256 buckets are then distributed among the disks in
    proportion to the size of each disk. The bucket distribution is done
    manually as part of the configuration (with the aid of an included
    tool, yars_generate_diskmap).

    The server is controlled with the command line tool yars.

    The basic operations of a running yars cluster are supporting requests
    of the form

      PUT http://$host/file/$filename
      GET http://$host/file/$md5/$filename
      HEAD http://$host/file/$md5/$filename
      GET http://$host/bucket_map

    to store and retrieve files, where $host may be any of the hosts in the
    cluster, $md5 is the md5 of the content, and $filename is a filename
    for the content to be stored. See Yars::Routes for documentation of
    other routes.

    Failover is handled in the following manner :

    If the host to which a file is assigned is not available, then the file
    will be "stashed" on the filesystem for the host to which it was sent.
    If there is no space there, other hosts and disks will be tried until
    an available one is found. Because of this failover mechanism, the
    "stash" must be checked whenever a GET request is handled. A successful
    GET will return quickly, but an unsuccessful one will take longer
    because all of the stashes on all of the servers must be checked before
    a "404 Not Found" is returned.

    Another tool yars_fast_balance is provided which takes files from
    stashes and returns them to their correct locations.

    A client Yars::Client is also available (in a separate distribution),
    for interacting with a yars server.

EXAMPLE 1

    The following sequence of commands will start yars on a single host
    (with 16 buckets) :

        $ mkdir ~/etc
        $ cat > ~/etc/Yars.conf
        ---
        start_mode : 'hypnotoad'
        url : http://localhost:9999
        hypnotoad :
          pid_file : /tmp/yars.pid
          listen :
             - http://localhost:9999
        servers :
        - url : http://localhost:9999
          disks :
            - root : /usr/local/data/disk1
              buckets : [ <%= join ',', '0'..'f' %> ]
        ^D
    
        $ yars start

    Now, verify that it works :

        $ GET http://localhost:9999/status

    And try to PUT and GET a file :

        echo "hi" | lwp-request -em PUT http://localhost:9999/file/here
        # (notice the "Location" header
        GET http://localhost:9999/file/764efa883dda1e11db47671c4a3bbd9e/here

    Also you can use Yars::Client :

        echo "hi" > myfile
        yarsclient upload myfile
        yarsclient download myfile 764efa883dda1e11db47671c4a3bbd9e

    Or to see the requests and responses :

        yarsclient --trace root upload myfile
        yarsclient --trace root download myfile 764efa883dda1e11db47671c4a3bbd9e

EXAMPLE 2

    To install Yars on a cluster of several hosts, the configuration for
    each host should be identical, except that the 'url' should reflect the
    host on which the server is running.

    To accomplish this, the above configuration may be divided into two
    files, one with the bucket map, and another with the server specific
    information.

        yars1 ~$ cat > ~/etc/Yars.conf :
        ----
        extends_config 'disk_map';
        url : http://yars1:9999
        hypnotoad :
          pid_file : /tmp/yars.pid
          listen :
             - http://yars1:9999
    
        yars2 ~$ cat > ~/etc/Yars.conf :
        ----
        extends_config 'disk_map';
        url : http://yars2:9999
        hypnotoad :
          pid_file : /tmp/yars.pid
          listen :
             - http://yars2:9999
    
        Then on both servers :
        $ cat > ~/etc/disk_map.conf :
        servers :
        - url : http://yars1:9999
          disks :
            - root : /usr/local/data/disk1
              buckets : [ <%= join ',', '0'..'9' %> ]
        - url : http://yars2:9999
          disks :
            - root : /usr/local/data/disk1
              buckets : [ <%= join ',', 'a'..'f' %> ]

    Then run "yars start" on both servers and voila, you have an archive.

    See also, clad, for a tool to facilitate running "yars start" on
    multiple hosts at once.

    Yars is the application package, it inherits from Clustericious::App
    and overrides the following methods :

 startup

    Called by the server to start up, we change the object classes to use
    Yars::Message::Request for incoming requests.

SEE ALSO

    Yars::Client

AUTHOR

    original author: Marty Brandon

    current maintainer: Graham Ollis <plicease@cpan.org>

    contributors:

    Brian Duggan

COPYRIGHT AND LICENSE

    This software is copyright (c) 2013 by NASA GSFC.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Yars

1.04      2014-12-21 01:07:35 -0500
  - remove bogus dep on Yars::Client

1.03      2014-12-19 12:05:50 -0500
  - merge Yars::Client into this distro
  - More mojo 5.x compat changes

1.02      2014-12-09 10:05:54 America/New_York
  - use Mojo::JSON's new API
    the older OO API was deprecated

1.01      2014-10-06 12:29:38 America/New_York
  - added --lock option for yars_disk_scan

1.00      2014-04-19 08:18:09 America/New_York
  - more robust checking for t/40_hostfail.t

0.99      2014-04-17 03:55:54 America/New_York
  - arguments to Yars::Tools->remote_stashed_server has changed
  - turn off IPv6 and SSL for t/40_hostfail.t
    see https://github.com/plicease/Test-Clustericious-Cluster/issues/3
    for details
  - make the content type be application/octet-stream 
    instead of text/plain for files with unrecognized or
    missing extensions

0.98      2014-04-16 03:19:23 America/New_York
  - rm dep on Smart::Comments
  - bump required Mojolicious to 4.31
  - Mojolicious 4.93 compatability

0.97_01   2014-04-08 11:16:00 America/New_York
  - add additional diagnostics for testing

0.97      2014-04-07 14:54:58 America/New_York
  - Fixed dynamic configuration bug introduced in
    version 0.96

0.96      2014-04-07 12:59:27 America/New_York
  - more portability fixes

0.95      2014-04-07 11:04:22 America/New_York
  - fix test failure on Solaris (GH#10)

0.94      2014-04-01 12:21:16 America/New_York
  - fix missing dependency (File::Find::Rule)
  - remove dependency on Try::Tiny

0.93      2014-03-28 13:47:55 America/New_York
  - yars_generate_diskmap: ignore blank lines and allow # style comments in input

0.92      2013-11-18 11:41:52 America/New_York
  - fix sloness of t/60_mark_down.t introduced in mojo 4.50

0.91_02   2013-11-07 07:10:44 America/New_York
  - fix yars_fast_balance.pm so that perl -c works whithout config

0.91_01   2013-11-06 11:50:32 America/New_York
  - /servers/status does more checking to see that each disk really is "up"
    This includes writing to the disks tmp directory and checking that the
    file is non zero size.  While this makes it slower, this route is usually
    used by Yars::Client manually by a human to check the health of the Yars
    cluster.

0.91      2013-10-06 10:55:59 America/New_York
  - more robust testing on platforms that don't support
    read only folders (cygwin)

0.90      2013-09-27 15:15:01 America/New_York
  - fixed false failure in test suite

0.89      2013-08-31 20:46:18 America/New_York
  - added generate_config for 'yars configure' command

0.88      2013-07-19 22:59:00 America/New_York
  - require Test::Clustericious::Cluster 0.08 for MSWin32 fixes

0.87      2013-07-18 09:23:59 America/New_York
  - promote to production

0.86_02   2013-07-13 12:58:27 America/New_York
  - basic use test for yars_fast_balance

0.86_01   2013-07-10 22:15:35 America/New_York
  - yars_generate_diskmap: dash character is legal for hostnames
  - yars_generate_diskmap: ports other than 9001
  - added test for yars_disk_scan
  - add a sanity check: our url has to be in the server list somewhere.

0.86      2013-07-10 17:01:59 America/New_York
  - added a test case for yars_generate_diskmap
  - documentation
  - fix $tools->_ua bug

0.85      2013-07-08 09:08:02 America/New_York
  - promote to production

0.84_02   2013-07-06 09:05:53 America/New_York
  - require Clustericious::Config 0.22

0.84_01   2013-07-05 22:33:00 America/New_York
  - Make OO interface to Yars::Tools

0.84      2013-07-02 17:23:54 America/New_York
  - Make Yars::Client 0.87 a prereq since it has fixes for Mojo 4.x

0.83_11   2013-07-01 19:02:52 America/New_York
  - improved Mojo 4.x detection and recovery

0.83_10   2013-07-01 07:26:15 America/New_York
  - remove dep on YAML (use YAML::XS instead which is already a prereq of Clustericious)

0.83_09   2013-06-30 18:15:57 America/New_York
  - retire problematic 071_tempdir test (replacement is 073_tempdir)

0.83_08   2013-06-28 11:35:40 America/New_York
  - fixed testing diagnostic

0.83_07   2013-06-26 21:38:13 America/New_York
  - added testing diagnostic

0.83_06   2013-06-26 15:22:36 America/New_York
  - added testing diagnostic

0.83_05   2013-06-22 22:59:29 America/New_York
  - use fewer resources during testing

0.83_04   2013-06-22 20:06:51 America/New_York
  - don't hardcode /tmp anymore

0.83      2013-05-25 22:04:54 America/New_York
  - Require Clustericious 0.9920 for Mojolicious 4.0 compatability

0.82      2013-05-24 04:47:14 America/New_York
  - Mojolicious 4.0 compatability

0.81      2013-05-06 08:31:53 EST5EDT
  - promote to production

0.80_06   2013-05-04 22:13:33 EST5EDT
  - Minor timing fixes

0.80_05   2013-05-04 18:26:46 EST5EDT
  - properly cleanup processes after test

0.80_04   2013-05-04 17:08:25 EST5EDT
  - fixed prereq
  - tests can now be run in parallel
  - tests should pass even if the perl in path != the build perl

0.80_03   2013-05-04 07:27:05 EST5EDT
  - Jump through some hoops so the right perl is executed in tests

0.80_02   2013-05-03 17:05:52 EST5EDT
  - Additional diagnostics

0.80_01   2013-05-03 16:34:52 EST5EDT
  - Less sleepy tests
  - use ephermeral ports for 020_dual.t to see if that plays nice
    with cpan testers
  - Portability: use File::HomeDir instead of $ENV{HOME}

0.80      2013-05-02 06:17:18 EST5EDT
  - move integration tests into the xt directory

0.79      2013-04-30 12:57:02 EST5EDT
     Remove extranious dep

0.78      2013-04-30 12:40:26 EST5EDT
     Compatability, bug check Mojolicious 3.84 (require 3.86)

0.77 2012-12-05
     Use /usr/bin/perl instead of /usr/bin/env perl
     for system tests to work.

0.76 2012-12-05
     More fixes to tests.

0.75 2012-11-29
     Improved portability of tests

0.74 2012-11-25
     First public release



lib/Yars/Client.pm
package Yars::Client;

# ABSTRACT: Yet Another RESTful-Archive Service Client
our $VERSION = '1.04'; # VERSION

use strict;
use warnings;

use Clustericious::Client;
use Clustericious::Client::Command;
use Clustericious::Config;
use Mojo::Asset::File;
use Mojo::ByteStream 'b';
use Mojo::URL;
use Mojo::Base '-base';
use File::Basename;
use File::Spec;
use Log::Log4perl qw(:easy);
use Digest::file qw/digest_file_hex/;
use Data::Dumper;
use IO::Uncompress::Gunzip qw(gunzip $GunzipError);
use Number::Bytes::Human qw( format_bytes parse_bytes );
use 5.10.0;

route_doc upload   => "<filename>";
route_doc content  => "<filename> <md5>";
route_doc download => "<filename> <md5> [dir]";
route_doc remove   => "<filename> <md5>";

has bucket_map_cached  => sub { 0; }; # Computed on demand.

route 'welcome'        => "GET",  '/';
route 'bucket_map'     => "GET",  '/bucket_map';
route 'disk_usage'     => "GET",  '/disk/usage';
route 'bucket_usage'   => "GET",  '/bucket/usage';
route 'servers_status' => "GET",  '/servers/status';
route 'get'            => "GET",  '/file', \"<md5> <filename>";
route 'check'          => "HEAD", '/file', \"<md5> <filename>";
route 'set_status'     => "POST", '/disk/status';
route 'check_files'    => "POST", '/check/manifest';

route_meta 'welcome'        => { auto_failover => 1, dont_read_files => 1 };
route_meta 'bucket_map'     => { auto_failover => 1, dont_read_files => 1 };
route_meta 'disk_usage'     => { auto_failover => 1, dont_read_files => 1 };
route_meta 'bucket_usage'   => { auto_failover => 1, dont_read_files => 1 };
route_meta 'servers_status' => { auto_failover => 1, dont_read_files => 1 };
route_meta 'get'            => { auto_failover => 1, dont_read_files => 1 };
route_meta 'check'          => { auto_failover => 1, dont_read_files => 1 };
route_meta 'set_status'     => { auto_failover => 1, dont_read_files => 1 };
route_meta 'check_files'    => { auto_failover => 1, dont_read_files => 1 };

route_meta 'upload'         => { dont_read_files => 1 };
route_meta 'download'       => { dont_read_files => 1 };
route_meta 'check_manifest' => { dont_read_files => 1 };
route_meta 'check'          => { dont_read_files => 1 };

route_args send => [
    { name => 'content',  type => '=s', required => 1 },
    { name => 'name',     type => '=s' },
];

route_args retrieve => [
    { name => 'location', type => '=s' },
    { name => 'name',     type => '=s' },
    { name => 'md5',      type => '=s' },
];

sub new {
    my $self = shift->SUPER::new(@_);
    $self->client->max_redirects(30);
    $self->client->connect_timeout(30);
    return $self;
}

sub _get_url {

    # Helper to create the Mojo URL objects
    my ($self, $path) = @_;

    my $url = Mojo::URL->new( $self->server_url );
    $url->path($path) if $path;

    return $url;
}

sub _hex2b64 {
    my $hex = shift or return;
    my $b64 = b(pack 'H*', $hex)->b64_encode;
    local $/="\n";
    chomp $b64;
    return $b64;
}

sub _b642hex {
    my $b64 = shift or return;
    # Mojo::Headers apparently become array refs sometimes
    $b64 = $b64->[0] if ref($b64) eq 'ARRAY';
    return unpack 'H*', b($b64)->b64_decode;
}

sub location {
    my ($self, $filename, $md5) = @_;

    ( $filename, $md5 ) = ( $md5, $filename ) if $filename =~ /^[0-9a-f]{32}$/i;
    LOGDIE "Can't compute location without filename" unless defined($filename);
    LOGDIE "Can't compute location without md5" unless $md5;
    $self->server_url($self->_server_for($md5));
    return $self->_get_url("/file/$md5/$filename")->to_abs->to_string;
}

sub download {
    # Downloads a file and saves it to disk.
    my $self = shift;
    my ( $filename, $md5, $dest_dir ) = @_;
    my $abs_url;
    if (@_ == 1) {
        $abs_url = shift;
        ($filename) = $abs_url =~ m|/([^/]+)$|;
    }
    ( $filename, $md5 ) = ( $md5, $filename ) if $filename =~ /^[0-9a-f]{32}$/i;

    if (!$md5 && !$abs_url) {
        LOGDIE "Need either an md5 or a url: download(url) or download(filename, md5, [dir] )";
    }

    my @hosts;
    @hosts  = $self->_all_hosts($self->_server_for($md5)) unless $abs_url;
    my $tries = 0;
    my $success = 0;
    my $host = 0;
    while ($tries++ < 10) {

        if ($tries > @hosts + 1) {
            TRACE "Attempt $tries";
            WARN "Waiting $tries seconds before retrying...";
            sleep $tries;
        }
        my $url;
        if ($abs_url) {
            $url = $abs_url;
        } else {
            $host = 0 if $host > $#hosts;
            $url = Mojo::URL->new($hosts[$host++]);
            $url->path("/file/$filename/$md5");
        }
        TRACE "GET $url";
        my $tx = $self->client->build_tx(GET => $url, { "Connection" => "Close", "Accept-Encoding" => "gzip" } );
        # TODO: set timeout for mojo 4.0
        $tx->res->max_message_size(parse_bytes($self->_config->max_message_size_client(default => 53687091200)));
        $self->client->start($tx);
        $self->res($tx->res);
        $self->tx($tx);
        my $res = $tx->success or do {
            my ($msg,$code) = $tx->error;
            if ($code) {
                ERROR "$code $msg";
                last;
            }
            if ($msg =~ /connection refused/i) {
                WARN "Error : $msg (may retry)";
                next;
            }
            WARN "Error (may retry) : $msg";
            next;
        };
        DEBUG "Received asset with size ".$res->content->asset->size;
        TRACE "Received headers : ".$res->headers->to_string;

        my $out_file = $dest_dir ? $dest_dir . "/$filename" : $filename;
        DEBUG "Writing to $out_file";
        if (my $e = $res->headers->header("Content-Encoding")) {
            LOGDIE "unsupported encoding" unless $e eq 'gzip';
            # This violate the spec (MD5s depend on transfer-encoding
            # not content-encoding, per
            # http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html
            # but we must support it.
            TRACE "unzipping $out_file";
            my $asset = $res->content->asset;
            gunzip($asset->is_file ? $asset->path : \( $asset->slurp )
                 => $out_file) or do {
                unlink $out_file;
                LOGDIE "Gunzip failed : $GunzipError";
            };
        } else {
            $res->content->asset->move_to($out_file);
        }
        my $verify = digest_file_hex($out_file,'MD5');
        $md5 ||= _b642hex($res->headers->header("Content-MD5"));

        unless ($md5) {
            WARN "No md5 in response header";
            next;
        }
        unless ($verify eq $md5) {
            WARN "Bad md5 for file (got $verify instead of $md5)";
            WARN "Response headers : ".$res->headers->to_string;
            unlink $out_file or WARN "couldn't remove $out_file : $!";
            WARN "Removed $out_file.  This is attempt $tries.";
            next;
        }

        $success = 1;
        last;
    }
    ERROR "Download failed." unless $success;
    return '' unless $success;
    return 'ok'; # return TRUE
}

sub remove {
    # Removes a file
    my ( $self, $filename, $md5 ) = @_;

    LOGDIE "file and md5 needed for remove"
        unless $filename && $md5;

    my $url = $self->_get_url("/file/$md5/$filename");
    TRACE("removing $filename $md5 from ", $url->to_string);

    # Delete the file
    $self->_doit(DELETE => $url);
}

# Given an md5, determine the correct server
# using a cached list of bucket->server assignments.
sub _server_for {
    my $self = shift;
    my $md5 = shift or LOGDIE "Missing argument md5";
    my $bucket_map = $self->bucket_map_cached;
    unless ($bucket_map && ref($bucket_map) eq 'HASH' && keys %$bucket_map > 0) {
        $bucket_map = $self->bucket_map or WARN $self->errorstring;
        $self->bucket_map_cached({ %$bucket_map }) if $bucket_map && ref $bucket_map && (keys %$bucket_map > 0);
    }
    unless ($bucket_map && ref $bucket_map && (keys %$bucket_map > 0)) {
        WARN "Failed to retrieve bucket map, using ".$self->server_url;
        return $self->server_url;
    }
    for (0..length($md5)) {
        my $prefix = substr($md5,0,$_);
        return $bucket_map->{ lc $prefix } if exists($bucket_map->{lc $prefix});
        return $bucket_map->{ uc $prefix } if exists($bucket_map->{uc $prefix});
    }
    LOGDIE "Can't find url for $md5 in bucket map : ".Dumper($bucket_map);
}

sub put {
    my $self = shift;
    my $remote_filename = shift;
    my $content = shift || join '', <STDIN>;
    # NB: slow for large content.
    my $md5 = b($content)->md5_sum;
    my $url = Mojo::URL->new($self->_server_for($md5));
    $url->path("/file/$remote_filename");
    TRACE "PUT $url";
    my $tx = $self->client->put("$url" => { "Content-MD5" => _hex2b64($md5), "Connection" => "Close" } => $content);
    $self->res($tx->res);
    $self->tx($tx);
    return $tx->success ? 'ok' : '';
}

sub _all_hosts {
    my $self = shift;
    my $assigned = shift;
    # Return all the hosts, any parameter will be put first in
    # the list.
    my @servers = ($assigned);
    push @servers, $self->server_url;
    push @servers, $self->_config->url;
    push @servers, @{ $self->_config->failover_urls(default => []) };
    my %seen;
    return grep { !$seen{$_}++ } @servers;
}

sub upload {
    my $self = shift;
    my $filename = pop;
    my $nostash;
    if (@_) {
        # To avoid failover :
        # yarsclient upload --nostash 1 foo
        # $yc->upload("--nostash",1","foo")
        # This is undocumented since it is only intended to be
        # used on a server when balancing, not as a public interface.

        if ($_[0] =~ /nostash$/) {
            shift;
            $nostash = shift;
        }
    }
    if (@_) {
        LOGDIE "unknown options to upload : @_";
    }

    LOGDIE "file needed for upload" unless $filename;
    $filename = File::Spec->rel2abs($filename);
    -r $filename or LOGDIE "Could not read " . $filename;

    # Don't read the file.
    my $basename = basename($filename);
    my $asset    = Mojo::Asset::File->new( path => $filename );
    my $md5      = digest_file_hex($filename, 'MD5');

    my @servers = $self->_all_hosts( $self->_server_for($md5) );

    my $tx;
    my $code;
    my $host;

    while (!$code && ($host = shift @servers)) {
        my $url = Mojo::URL->new($host);
        $url->path("/file/$basename/$md5");
        DEBUG "Sending $md5 to $url";

        my @nostash = ($nostash ? ("X-Yars-NoStash" => 1) : ());
        $tx = $self->client->build_tx(
            PUT => "$url" => {
                @nostash,
                "Content-MD5" => _hex2b64($md5),
                "Connection"  => "Close"
            }
        );
        $tx->req->content->asset($asset);
        # TODO: set timeout for mojo 4.0
        $tx = $self->client->start($tx);
        $code = $tx->res->code;
        $self->res($tx->res);
        $self->tx($tx);

        if (!$tx->success) {
            my ($msg,$code) = $tx->error;
            if ($code) {
                INFO "Failed to PUT to $host : $code $msg";
            } else {
                INFO "PUT to $host failed : ".$msg;
            }
        }
    }
    return '' if !$code || !$tx->res->is_status_class(200);

    DEBUG "Response : ".$tx->res->code." ".$tx->res->message;
    $self->res($tx->res);
    return 'ok';
}

sub _rand_filename {
    my $a = '';
    $a .= ('a'..'z','A'..'Z',0..9)[rand 62] for 1..43;
    return $a;
}

sub send {
    my $self = shift;
    my %args = $self->meta_for->process_args(@_);
    my $content = $args{content};
    my $filename = $args{name} || $self->_rand_filename;
    my $status = $self->put($filename, $content);
    return unless $status eq 'ok';
    return $self->res->headers->location;
}

sub retrieve {
    my $self = shift;
    my %args = $self->meta_for->process_args(@_);
    if (my $location = $args{location}) {
        my $tx = $self->client->get($location);
        my $res = $tx->success or do {
            $self->tx($tx);
            $self->res($tx->res);
            return;
        };
        return $res->body;
    }
    my $md5 = $args{md5} or LOGDIE "need md5 or location to retrieve";
    my $name = $args{name} or LOGDIE "need name or location to retrieve";
    return $self->get($md5,$name);
}

sub res_md5 {
    my $self = shift;
    my $res = $self->res or return;
    if (my $b64 = $res->headers->header("Content-MD5")) {
        return _b642hex($b64);
    }
    if (my $location = $res->headers->location) {
        my ($md5) = $location =~ m[/file/([0-9a-f]{32})/];
        return $md5;
    }
    return;
}

sub check_manifest {
    my $self     = shift;
    my @args     = @_;
    my $check    = 0;
    my $params  = "";
    my $manifest;
    while ($_ = shift @_) {
        /^-c$/ and do { $check = 1; next; };
        /^--show_corrupt$/ and do { $params = "?show_corrupt=" . shift; next; };
        $manifest = $_;
    }
    LOGDIE "Missing manifest" unless $manifest;
    LOGDIE "Cannot open manifest $manifest" unless -e $manifest;
    my $contents = Mojo::Asset::File->new(path => $manifest)->slurp;
    my $got      = $self->_doit(POST => "/check/manifest$params", { manifest => $contents  });
    return $got unless $self->tx->success;
    $got->{$manifest} = (@{$got->{missing}}==0 ? 'ok' : 'not ok');
    return { $manifest => $got->{$manifest} } if $check;
    return $got;
}

sub remote {
    my $self = shift;
    $self->bucket_map_cached(0);
    $self->SUPER::remote(@_);
}

1;



=pod

=head1 NAME

Yars::Client - Yet Another RESTful-Archive Service Client

=head1 VERSION

version 1.04

=head1 SYNOPSIS

 my $r = Yars::Client->new;

 # Send and retrieve content.
 my $location = $y->send(content => 'hello, world') or die $y->errorstring;
 say $y->retrieve(location => $location);

 # Alternatively, use names and md5s explicitly.
 my $location = $y->send(content => 'hello there', name => "greeting");
 my $md5 = $y->res_md5;
 say $y->retrieve(filename => 'greeting', md5 => $md5);

 # Upload a file.
 $r->upload($filename) or die $r->errorstring;
 print $r->res->headers->location;

 # Download a file.
 $r->download($filename, $md5) or die $r->errorstring;
 $r->download($filename, $md5, '/tmp');   # download it to the /tmp directory
 $r->download("http://yars/0123456890abc/filename.txt"); # Write filename.txt to current directory.

 # More concise version of retrieve.
 my $content = $r->get($filename,$md5);

 # Delete a file.
 $r->remove($filename, $md5) or die $r->errorstring;

 # Compute the URL of a file based on the md5 and the buckets.
 print $r->location($filename, $md5);

 print "Server version is ".$r->status->{server_version};
 my $usage = $r->disk_usage();      # Returns usage for a single server.
 my $nother_usage = Yars::Client->new(url => "http://host1:9999")->disk_usage();
 my $status = $r->servers_status(); # return a hash of servers, disks, and their statuses

 # Mark a disk down.
 my $ok = $r->set_status({ root => "/acps/disk/one", state => "down" });
 my $ok = $r->set_status({ root => "/acps/disk/one", state => "down", host => "http://host2" });

 # Mark a disk up.
 my $ok = $r->set_status({ root => "/acps/disk/one", state => "up" });

 # Check a manifest file or list of files.
 my $details = $r->check_manifest( $filename );
 my $check = $r->check_manifest( "-c", $filename );
 my $check = $r->check_manifest( "--show_corrupt" => 1, $filename );
 my $ck = $r->check_files({ files => [
     { filename => $f1, md5 => $m1 },
     { filename => $f2, md5 => $m2 } ] });

=head1 DESCRIPTION

Client for L<Yars>.

=head1 SEE ALSO

L<yarsclient>, L<Clustericious::Client>

=cut

=head1 AUTHOR

original author: Marty Brandon

current maintainer: Graham Ollis <plicease@cpan.org>

contributors:

Brian Duggan

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by NASA GSFC.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__

